cdap-ui/app/cdap/components/DAG/DAGProvider.tsx

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright © 2018 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+*/
+
+import * as React from 'react';
+import { List, Map } from 'immutable';
+
+export interface INodeConfig {
+  label?: string;
+  styles?: object;
+  properties?: object;
+  [newProps: string]: any;
+}
+
+export interface INode {
+  id: string;
+  name: string;
+  config: INodeConfig;
+}
+export interface IConnection {
+  [extraProps: string]: any;
+  source?: any;
+  target?: any;
+  sourceId: string;
+  targetId: string;
+}
+export interface IAdjacencyMap {
+  [key: string]: string;
+}
+
+export interface IDagProviderState {
+  adjancecyMap: Map<string, string | object>;
+  connections: List<IConnection>;
+  nodes: List<INode>;
+}
+export interface IDagStore extends IDagProviderState {
+  addNode: (node: INode) => void;
+  addConnection: (connection: IConnection) => void;
+  getConnections: () => List<IConnection>;
+  getNodes: () => List<INode>;
+  removeNode: (node: string) => void;
+  removeConnection: (connectionObj: IConnection) => void;
+}
+
+export const MyContext = React.createContext({} as Readonly<IDagStore>);
+
+export class DAGProvider extends React.Component<any, IDagProviderState> {
+  private getNodes = (): List<INode> => this.state.nodes;
+  private addNode = (node: INode) =>
+    this.setState({
+      nodes: this.state.nodes.push(node),
+    });
+  private removeNode = (nodeId: string) =>
+    this.setState({
+      nodes: this.state.nodes.filter((node) => node.id !== nodeId) as List<INode>,
+    });
+  private getConnections = (): List<IConnection> => this.state.connections;
+  private addConnection = (connection: IConnection) => {
+    if (
+      this.state.connections.find(
+        (conn) => conn.sourceId === connection.sourceId && conn.targetId === connection.targetId
+      )
+    ) {
+      return;
+    }
+    this.setState({
+      connections: this.state.connections.push(connection),
+    });
+  };
+  private removeConnection = (connectionObj: IConnection) => {
+    const newConnections = this.state.connections.filter(
+      (connection: IConnection) =>
+        connection.from === connectionObj.from && connection.to === connectionObj.to
```

Comment:
Should this be !== for removing the connectionObj?

Reply:
Good catch.